### ISSUE TITLE
Handler.post() actually does a postAtFrontOfQueue() if currentTime == 0

### ISSUE THREAD

**Issue by kriegfrj** [robolectric/robolectric#1740]
This is an artefact of the changes that I made in #1466 & #1431. In this change, I removed a lot of the queuing code in favour of using the real Android code for `MessageQueue` and `Handler`, with a bit of glue to hook it into the `Scheduler`. 

Normally `MessageQueue.enqueueMessage()` queues messages to run after all previously queued messages with the same `when` parameter - eg, if you queue msg1 to run at 100ms and later queue msg2 to run at 100ms, then when the clock gets to 100ms msg1 will run first followed immediately by msg2. However, `enqueueMessage()` treats the case where `when` = 0 specially, and always adds it to the head of the queue. This special-case code is used by `Handler.sendMessageAtFrontOfQueue()` (and `postAtFrontOfQueue()`, which depends on this method) to implement the "post at front of queue" functionality, by internally enqueuing the message with a `when` of 0ms.

In the real world this special-casing doesn't cause any problems, because the real time on the system clock will always be > 0ms. However, in Robolectric's simulated world, `Scheduler`s are always initialized with 0ms on the clock. So if you do `Handler.post(r)` or `Handler.postDelayed(r, 0)` on a newly-initialized scheduler, it ends up behaving the same as `Handler.postAtFrontOfQueue(r)`. For example:

````
Handler h = new Handler();
Robolectric.getForegroundScheduler().pause();
h.post(new Runnable() { public void run() { System.out.println("one"); } } );
h.post(new Runnable() { public void run() { System.out.println("two"); } } );
Robolectric.getForegroundScheduler().unPause();
````
... this will print "two, one" instead of "one, two" as you'd expect in real Android. What would be even more confusing is that if you advance the clock on the scheduler (eg, by 1ms) and do the same thing, you'll get the expected output of "one, two".

I have been aware of it for a while but I didn't appreciate that it might cause real problems until I came across #919. I started to wonder if there are unexplained issues that might have this bug as their root cause (maybe #1587?).

**Comment by kriegfrj** [comment]
Trying to think of the best way to fix this.

The simplest fix would simply be to have `Scheduler` initialize with a start time of something greater than 0ms. Arguably this makes it closer to real-world Android; however this may break a lot of existing tests.

The only other way I can see to fix it requires completely re-implementing `MessageQueue.enqueueMessage()` - a re-implementation that would be 99% a copy of the real Android code. I was hoping to avoid this, but I can't see any other way.

**Comment by jongerrish** [comment]
I would vote for #1 - feels the most natural.

I'm not sure the impact is as big as you think, but we have a large corpus
of tests and I'm happy to run any potential change against them if you want
a damage assessment :-)

**Comment by kriegfrj** [comment]
Feel free to give it a try - it would be useful to have that reference. It should be as simple as initializing `Scheduler.currentTime` to some positive integer (1 would probably work). I guess the impact would be most likely in tests that measure timing, and I understand that most people's unit tests probably don't mess with timing. However, I know that there is at least one assertion in `ShadowMediaPlayerTest` that will fail with this change.

**Comment by kriegfrj** [comment]
Closed via #1839.
